Thermal Comfort Mean Radiant Temperature Calculation Upgrades

RKStrand opened the pull request:
Pull request overview
---------------------
This pull request addresses issues that first came up in GitHub Issue #3469.  This started as a documentation fix and then became a more critical look at how we are calculating various flavors of MRT in the thermal comfort routine.  Two were identified as having something less than ideal.  The surface weighted option which averages one surface temperature and the zone MRT was basically double counting the surface that the person was closest to.  So, this was altered so that the zone MRT used for this calculation did not include the effect of the surface the person was closest to.  The angle factor option was using a simplified equation that did not account for surface emissivity and was also using a linear version of the equation.  Both of those deficiencies were addressed with this work.  Two unit tests and documentation mods were also completed.

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue #3469 

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Code style (parentheses padding, variable names)
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check -- verify this
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add rules to spreadsheet
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Documentation changes in place
 - [ ] Changed docs build successfully
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

RKStrand committed:
Changes to implement corrections

Changes made to code to fix the surface weighted MRT calculation to not
double count the surface that is being weighted and the calculation of
the angle factor based MRT.

RKStrand committed:
Correction of bugs in fix

Corrected a mistake in the bug fix for the surface weighted option.

RKStrand committed:
minor mod made during testing

Uses different variable for surface temperature though technically they
are identical. Tested to verify that differences were due to removing
one surface from MRT calculation.  When SurfNum2 != SurfNum condition
was removed from two IF blocks, results were identical to develop for
input files that use SurfaceWeighted.

RKStrand committed:
Addition of two unit tests

Two unit tests were added to exercise the new Real64 functions for the
modification of the surface weighted MRT and angle factor based MRT
calculations.

RKStrand committed:
Modification of engineering reference document

Modifications were made to the engineering reference so that the new
upgrades in the code are now properly described in this document.

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into 121865809-MRT-Thermal-Comfort-Revisions

mjwitte commented on src/EnergyPlus/ThermalComfort.cc:
```diff
@@ -1881,47 +1708,51 @@ namespace ThermalComfort {
+				thisAngFacList.SurfacePtr( SurfNum ) = FindItemInList( cAlphaArgs( SurfNum + 2 ), Surface );
+				thisAngFacList.AngleFactor( SurfNum ) = rNumericArgs( SurfNum );
+				thisAngFacList.SurfaceEmissivity( SurfNum ) = Construct( Surface( thisAngFacList.SurfacePtr( SurfNum ) ).Construction ).InsideAbsorpThermal;
+				thisAngFacList.SumSurfaceEmissAngleFactor += thisAngFacList.SurfaceEmissivity( SurfNum ) * thisAngFacList.AngleFactor( SurfNum );
```
Nice efficiency idea to precalculate this, but `InsideAbsorpThermal` can be changed with EMS, so this needs to be recalculated every timestep just in case.

RKStrand replied:
@mjwitte Ah, I didn't realize that--good catch.  So much for efficiency when you have EMS!  I will fix this.

mjwitte commented on src/EnergyPlus/ThermalComfort.cc:
```diff
@@ -1985,35 +1807,94 @@ namespace ThermalComfort {
+	)
+	{
+		
+		// Purpose: Calculate a modified zone MRT that excludes the Surface( SurfNum ).
```
This function does more than this, it gets the final answer including the weighted surface.

RKStrand replied:
@mjwitte Good point.  I will correct this.

mjwitte commented on src/EnergyPlus/ThermalComfort.cc:
```diff
@@ -1985,35 +1807,94 @@ namespace ThermalComfort {
+			CalcSurfaceWeightedMRT = 0.5 * ( TH( 2, 1, SurfNum ) + ( SumAET / ZoneAESum( ZoneNum ) ) );
+		} else {
+			ShowWarningError( "Zone areas*inside surface emissivities are summing to zero, for Zone=\"" + Zone( ZoneNum ).Name + "\"" );
+			ShowContinueError( "As a result, MAT will be used for MRT when calculating a surface weighted MRT for this zone." );
```
Don't want these warnings to write every timestep.  Sufficient to write this once - and add "only the first occurrence is reported" or something to that effect.

RKStrand replied:
@mjwitte Ok, I will change this as well.  It shouldn't ever happen, but I get that if it does that no one wants a 20Gb file with this message repeated over and over and over again.

mjwitte commented on src/EnergyPlus/ThermalComfort.cc:
```diff
@@ -1985,35 +1807,94 @@ namespace ThermalComfort {
+			ZoneAESum = 0.0;
+			for ( SurfNum2 = 1; SurfNum2 <= TotSurfaces; ++SurfNum2 ) {
+				if ( Surface( SurfNum2 ).HeatTransSurf ) {
+					SurfaceAE( SurfNum2 ) = Surface( SurfNum2 ).Area * Construct( Surface( SurfNum2 ).Construction ).InsideAbsorpThermal;
```
Again, this can't be precalculated if EMS actuates the thermal absorptance.  Since this is acting on a single zone now, it should simple enough to just have one loop that calculates both ZoneAESum and SumAET, exlcuding the weighted surface.

RKStrand replied:
@mjwitte Got it.  Will fix this also.

mjwitte reviewed (commented):
Some initial comments.

mjwitte commented:
@RKStrand Also, I merged in latest develop on this branch, so be sure to pull before making changes.

Myoldmopar commented:
@RKStrand as soon as you get his comments addressed I'll make one final pass over this and if all is well I'll get it merged in.

mjwitte commented on doc/engineering-reference/src/simulation-models-encyclopedic-reference-005/occupant-thermal-comfort.tex:
```diff
@@ -1041,66 +1041,66 @@ \subsubsection{References}\label{references-033}
+T\(_{r}\) & Mean radiant temperature & °C & - \tabularnewline
+T\(_{r-avg}\) & Zone averaged radiant temperature & °C & -  \tabularnewline
+T\(_{surf}\) & Surface temperature & °C & -  \tabularnewline
+F\(_{surf}\) & Angle factor between person and surface & - & 0-1  \tabularnewline
```
Minor detail, the nomenclature table is missing A and epsilon.

RKStrand committed:
Addressed review comments

This commit hopefully corrects all of the issues noted during the
review.  Results should be similar/same to the commit that was done
just prior to the pull request.

RKStrand commented:
@mjwitte @Myoldmopar Ok, I think I have addressed all of Mike's comments and have pushed another commit up.  Once the CI cycle completes I will take one more look at the results but I think this should be ready to go.  Thanks again for your help!

Myoldmopar commented:
I'll let the debug builds run before merging, but this seems good now.

Myoldmopar committed:
Remove unused using statement causing the build warning

Myoldmopar committed:
Fix doc build error

Myoldmopar commented:
Fixed both the build warning, and the Linux build error (which was due to failing documentation build).  Not waiting for another pass of CI, this should be good.  Merging.
